### PR TITLE
[Opt] Build a control-flow graph for optimization

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #926 #656

This will serve as the first stage of refactoring `irpass::variable_optimization` to data flow analysis & optimization.

This PR includes unreachable code elimination by the CFG.

Benchmark:
```
copy_from::scalar_________________________________________
codegen_statements            ?[35m     35?[39m -> ?[36m     27 ?[32m   -22.9%
?[39m
codegen_kernel_statements     ?[35m     35?[39m -> ?[36m     27 ?[32m   -22.9%
?[39m
```
Case: `continue::unconditional_continue` (the test before copy_from::scalar -- the benchmark has the tests shifted by one)
**Before**:
```
kernel {
  $0 = offloaded range_for(0, 1) block_dim=adaptive {
    <i32 x1> $1 = const [0]
    <i32 x1> $2 = const [1000]
    $3 : for in range($1, $2, step 1) {
      <i32 x1> $4 = loop $3 index 0
      $5 continue (scope=$5)
      <gen*x1> $6 = get root
      <gen*x1> $7 = [S0root][root]::lookup($6, $1) activate = false
      <gen*x1> $8 = get child [S0root->S1dense] $7
      <i32 x1> $9 = bit_extract($4 + 0, 0~10)
      <gen*x1> $10 = [S1dense][dense]::lookup($8, $9) activate = false
      <i32*x1> $11 = get child [S1dense->S2place_i32] $10
      <i32*x1> $12 : global store [$11 <- $4]
    }
  }
}
```
**After**:
```
kernel {
  $0 = offloaded range_for(0, 1) block_dim=adaptive {
    <i32 x1> $1 = const [0]
    <i32 x1> $2 = const [1000]
    $3 : for in range($1, $2, step 1) {
      $4 continue (scope=$4)
    }
  }
}
```
benchmark20200527

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by xumingkuan** [comment]
For `serial` offloads, we can build a `CFGNode` (basic block in the control-flow graph) for each offload if it only consists of a basic block, and create an edge from the previous offload to the next offload.

For `struct_for` offloads, the body may not be executed. I think we can introduce some empty nodes (which is easy to implement) like
```
empty node 1
|        \
|  struct_for offload node 2
|        /
empty node 3
      |
serial offload node 4
```

For offloads without a body, we can just treat them as a normal statement (and probably ignore them).

(To conclude the above 3 cases: create a `CFGNode` between every two offloads with bodies. The node may be empty or contain offloads without a body.)

For `range_for` offloads, however, I'm not sure how to deal with its implicit loading `GlobalTemp[begin_offset]` if `const_begin` is `false` (and same about `end_offset`).

I want to deal with the offloaded `range_for` and the `for` loops inside `OffloadedStmt` in a unified way, without storing too much stuff in `CFGNode`.
@yuanming-hu Do you have any ideas?

**Comment by yuanming-hu** [comment]
Oh, actually after offloading, we only need to build the CFG within `OffloadedStmt`.
Optimization across `OfflodedStmt` will be done via the async engine :-)

**Comment by xumingkuan** [comment]
About the loops inside offloads, shall we store an `std::vector<Stmt *> loop_vars` in each `CFGNode`?

Or shall we use `LoopIndexStmt` instead of `alloca`, which will make `CFGNode` neater, but will it cause problems when there are nested loops?

**Comment by yuanming-hu** [comment]
`std::vector<Stmt *> loop_vars` is actually a bad design and we should try to gradually get rid of it...

How about adding a `Stmt *loop` member to `LoopIndexStmt` so that it works for nested loops? `loop` can be a `RangeForStmt`, `StructForStmt`, or `OffloadedStmt`.

**Comment by xumingkuan** [comment]
> How about adding a `Stmt *loop` member to `LoopIndexStmt` so that it works for nested loops? `loop` can be a `RangeForStmt`, `StructForStmt`, or `OffloadedStmt`.

Sounds good. I'll do this refactor as the first thing next week. There's one thing I'm slightly confused: why isn't there `void CodeGenLLVM::visit(StructForStmt *for_stmt)` in codegen_llvm.cpp? Does it mean that struct for can only appear in the top-level?

**Comment by yuanming-hu** [comment]
Yes, struct for is pretty complex (needs list generation etc) and can only appear at the top level.

**Comment by archibate** [comment]
@taichi-gardener commented now:

Hello, I'm a diligent gardener powered by GitHub &copy; Actions.

**I noticed that this PR seems to be staled?**

>> noreply@github.com from github-actions[bot]

**Comment by xumingkuan** [comment]
> @taichi-gardener commented now:
> 
> Hello, I'm a diligent gardener powered by GitHub © Actions.
> 
> **I noticed that this PR seems to be staled?**
> 
> > > [noreply@github.com](mailto:noreply@github.com) from github-actions[bot]

This PR is of a heavy workload, and I'm working on removing `loop_vars` before continue working on this PR...

**Comment by xumingkuan** [comment]
https://github.com/taichi-dev/taichi/blob/29eae0bd7220f49329a884741ca2fa1da096255a/taichi/ir/statements.cpp#L22-L24

I found this confusing. I thought `stmt->has_body()` is equivalent to `stmt->body` in an `if` condition...

**Comment by xumingkuan** [comment]
I assumed the `IRNode *root` is a `Block *` in `std::unique_ptr<ControlFlowGraph> build_cfg(IRNode *root)`, but it may also work with `OffloadedStmt *` with `has_body() == true`.

**Comment by yuanming-hu** [comment]
Sorry, but I won't have time for reviewing this until Friday. Please work on something orthogonal so that you are not blocked. Thanks!

**Comment by archibate** [comment]
tl;dr yet, a name nit: `cfg` make people think of `configuration`, how about:
`CtrlflowGraph` and `CGBuilder`? Although `cg` make people think of `computer graphics` too.. sorry about being picky here.

**Comment by xumingkuan** [comment]
> tl;dr yet, a name nit: `cfg` make people think of `configuration`, how about:
> `CtrlflowGraph` and `CGBuilder`? Although `cg` make people think of `computer graphics` too.. sorry about being picky here.

Good point. Actually, besides "configuration" and "control-flow graph", CFG also means "context-free grammar"...

"cg" sounds weird to me. I'm already using `ControlFlowGraph` as the class name, and `control_flow_graph_optimization` as the function name seems too long...

File: taichi/transforms/cfg_optimization.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#include "taichi/ir/cfg.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/analysis.h"
+
+TLANG_NAMESPACE_BEGIN
+
+namespace irpass {
+void cfg_optimization(IRNode *root) {
+  auto cfg = analysis::build_cfg(root);
+  while (cfg->unreachable_code_elimination());
```
[the remark this answers is not in the document]
Comment (xumingkuan) [reply taichi/transforms/cfg_optimization.cpp:11]: See #1059 -- we don't want to introduce exceptions unless there are no better solutions.

However, I admit this line is a bit confusing. I'll make it more readable.

File: taichi/analysis/build_cfg.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    prev_nodes.clear();
+    begin_location = next_begin_location;
+    last_node_in_current_block = node;
+    return node;
+  }
+
+  void visit(ContinueStmt *stmt) override {
+    // Don't put ContinueStmt in any nodes.
+    continues_in_current_loop.push_back(new_node(current_stmt_id + 1));
```
Comment (yuanming-hu) [inline taichi/analysis/build_cfg.cpp:51]: Slightly confused about this: what is this `new_node` for, if not `continue`?

Comment (xumingkuan) [reply taichi/analysis/build_cfg.cpp:51]: For example,
```
a = 1
b = 1
continue
c = 1
```
`new_node` will construct a node with
```
a = 1
b = 1
```
and the next node will begin at `current_stmt_id + 1`, that is,
```
c = 1
```
